### PR TITLE
Fix performance of 2d*2d np.dot() on vector arguments

### PR DESCRIPTION
**Pull request by pitrou**
While it is common to represent vectors as 1-d Numpy arrays, some people prefer the more mathematically correct 2-d representation (either `(n, 1) or (1, n)`). Numba considered np.dot() with such arguments as a regular matrix * matrix product, which is pessimal. This PR selects the optimal implementation at runtime, to fix the performance discrepancy with Numpy.

### CODE REVIEW

File: numba/tests/test_linalg.py [hunk 1]
```diff
@@ -167,18 +166,28 @@ def samples(m, n, k):
...
-            self.check_func(pyfunc2, cfunc2, (a, b))
-            self.check_func(pyfunc2, cfunc2, (b.T, a.T))
         if pyfunc3 is not None:
             cfunc3 = jit(nopython=True)(pyfunc3)
+
+        # Test generic matrix * matrix as well as "degenerate" cases
+        # where one of the outer dimensions is 1 (i.e. really represents
+        # a vector, which may select a different implementation)
+        for m, n, k in [(2, 3, 4),  # Generic matrix * matrix
+                        (1, 3, 4),  # 2d vector * matrix
```
Comment (stuartarchibald) [inline numba/tests/test_linalg.py]: Is it perhaps worth doing a matrix * 2d vector too ?

File: numba/targets/arraymath.py [hunk 2]
```diff
@@ -894,47 +971,42 @@ def check_args(a, b, out):
...
-                            builder.bitcast(alpha, ll_void_p), data_a, lda,
-                            data_b, ldb, builder.bitcast(beta, ll_void_p),
-                            data_c, ldc))
-    check_blas_return(context, builder, res)
+    x_data = x.data
+    y_data = y.data
+    out_data = out.data
+
+    # Check whether any of the operands is really a 1-d vector represented
+    # as a (1, k) or (k, 1) 2-d array.  In those cases, it is pessimal
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Is it worth considering (1,1,....,k) and (k,1,1.....1) too ?

File: numba/tests/test_linalg.py (hunk 1 above)
Comment (pitrou) [inline numba/tests/test_linalg.py]: It's done below, when swapping the arguments and transposing them.

File: numba/targets/arraymath.py (hunk 2 above)
Comment (pitrou) [inline numba/targets/arraymath.py]: More than 2d arrays are not supported currently.

File: numba/tests/test_linalg.py (hunk 1 above)
Comment (stuartarchibald) [inline numba/tests/test_linalg.py]: Ah yes, thanks.

File: numba/targets/arraymath.py (hunk 2 above)
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Fair enough. I guess in general use most people are either in the "vectors are orientation-less data containers" or "vectors are 2D matrix specialisation" category.

**Comment by stuartarchibald** [comment]
This code ensures that a missed optimisation opportunity is now performed. The sample code that displayed the performance problem is now considerably faster and running at the expected speed against this branch on my local machine. The additional test coverage seems adequate for the additions made.

Many thanks for the patches. +1 from me.

**Pull request merged**